Error recording retrieval failures

I'm now seeing the following error at the end of most of my lassie fetches. Not sure if the error is always related to HTTP or not.
```
2023-07-17T11:47:49.852-0700    ERROR   lassie/retriever        retriever/parallelpeerretriever.go:356  Error recording retrieval failure for protocol transport-ipfs-gateway-http: internal error, no such active retrieval 3145cc82-5a58-4866-9580-0553ba85effb
```

Here is one example of the fetch progress logs I was running
```
$ ./lassie fetch -o - -p bafybeic56z3yccnla3cutmvqsn5zy3g24muupcsjtoyp3pu5pm5amurjx4 | car extract

Fetching bafybeic56z3yccnla3cutmvqsn5zy3g24muupcsjtoyp3pu5pm5amurjx4
Querying indexer for bafybeic56z3yccnla3cutmvqsn5zy3g24muupcsjtoyp3pu5pm5amurjx4...
Found 12 storage provider candidate(s) in the indexer:
        12D3KooWPgBdZSbmKbD7ZQGjU7gZCcCKGvWSnBf1q4xAbpDdtJaJ, Protocols: [transport-ipfs-gateway-http]
        12D3KooWKGCcFVSAUXxe7YP62wiwsBvpCmMomnNauJCA67XbmHYj, Protocols: [transport-graphsync-filecoinv1]
        12D3KooWKGCcFVSAUXxe7YP62wiwsBvpCmMomnNauJCA67XbmHYj, Protocols: [transport-bitswap]
        QmQzqxhK82kAmKvARFZSkUVS6fo9sySaiogAnx5EnZ6ZmC, Protocols: [transport-bitswap]
        QmUA9D3H7HeCYsirB3KmPSvZh3dNXMZas6Lwgr4fv1HTTp, Protocols: [transport-ipfs-gateway-http]
        QmQzqxhK82kAmKvARFZSkUVS6fo9sySaiogAnx5EnZ6ZmC, Protocols: [transport-bitswap]
        QmUA9D3H7HeCYsirB3KmPSvZh3dNXMZas6Lwgr4fv1HTTp, Protocols: [transport-ipfs-gateway-http]
        12D3KooWJ8YAF6DiRxrzcxoeUVjSANYxyxU55ruFgNvQB4EHibpG, Protocols: [transport-graphsync-filecoinv1]
        12D3KooWJqxfjvbDkKByJQeT5urmeFBavNhNvidsAeqjWr6hf7jT, Protocols: [transport-bitswap]
        12D3KooWDCXxiSsLi1NT9tsiyimwV6YstQkrjTjD2hAkz2KRVAGG, Protocols: [transport-ipfs-gateway-http]
        12D3KooWJWAwD6CbJnDm6tHcFuZxvWrjhp53hDjLHM1GiP1pYXmY, Protocols: [transport-graphsync-filecoinv1]
        12D3KooWJWAwD6CbJnDm6tHcFuZxvWrjhp53hDjLHM1GiP1pYXmY, Protocols: [transport-bitswap]
Retrieving from [Bitswap] (started-retrieval)...
Retrieving from [12D3KooWDCXxiSsLi1NT9tsiyimwV6YstQkrjTjD2hAkz2KRVAGG] (started-retrieval)...
Retrieving from [12D3KooWDCXxiSsLi1NT9tsiyimwV6YstQkrjTjD2hAkz2KRVAGG] (connected-to-provider)...
Retrieving from [12D3KooWJWAwD6CbJnDm6tHcFuZxvWrjhp53hDjLHM1GiP1pYXmY] (started-retrieval)...
Retrieving from [QmUA9D3H7HeCYsirB3KmPSvZh3dNXMZas6Lwgr4fv1HTTp] (started-retrieval)...
Retrieving from [12D3KooWPgBdZSbmKbD7ZQGjU7gZCcCKGvWSnBf1q4xAbpDdtJaJ] (started-retrieval)...
Retrieving from [QmUA9D3H7HeCYsirB3KmPSvZh3dNXMZas6Lwgr4fv1HTTp] (connected-to-provider)...
Retrieving from [12D3KooWKGCcFVSAUXxe7YP62wiwsBvpCmMomnNauJCA67XbmHYj] (started-retrieval)...
Retrieving from [12D3KooWJ8YAF6DiRxrzcxoeUVjSANYxyxU55ruFgNvQB4EHibpG] (started-retrieval)...
Retrieving from [12D3KooWPgBdZSbmKbD7ZQGjU7gZCcCKGvWSnBf1q4xAbpDdtJaJ] (connected-to-provider)...
Retrieving from [12D3KooWJWAwD6CbJnDm6tHcFuZxvWrjhp53hDjLHM1GiP1pYXmY] (connected-to-provider)...
Retrieving from [12D3KooWJWAwD6CbJnDm6tHcFuZxvWrjhp53hDjLHM1GiP1pYXmY] (proposed)...
Retrieving from [12D3KooWJ8YAF6DiRxrzcxoeUVjSANYxyxU55ruFgNvQB4EHibpG] (connected-to-provider)...
Retrieving from [12D3KooWKGCcFVSAUXxe7YP62wiwsBvpCmMomnNauJCA67XbmHYj] (connected-to-provider)...
Retrieving from [Bitswap] (first-byte-received)...
Received 18 blocks / 15 MiB....
Fetched [bafybeic56z3yccnla3cutmvqsn5zy3g24muupcsjtoyp3pu5pm5amurjx4] from []:
        Duration: 7.760195108s
          Blocks: 18
           Bytes: 15 MiB
2023-07-17T11:47:49.852-0700    ERROR   lassie/retriever        retriever/parallelpeerretriever.go:356  Error recording retrieval failure for protocol transport-ipfs-gateway-http: internal error, no such active retrieval 3145cc82-5a58-4866-9580-0553ba85effb
2023-07-17T11:47:49.853-0700    ERROR   lassie/retriever        retriever/parallelpeerretriever.go:356  Error recording retrieval failure for protocol transport-ipfs-gateway-http: internal error, no such active retrieval 3145cc82-5a58-4866-9580-0553ba85effb
2023-07-17T11:47:49.853-0700    ERROR   lassie/retriever        retriever/parallelpeerretriever.go:356  Error recording retrieval failure for protocol transport-ipfs-gateway-http: internal error, no such active retrieval 3145cc82-5a58-4866-9580-0553ba85effb
extracted 1 file(s)
```

It connects to three HTTP providers, gets it's first byte via Bitswap, and then I get the three errors at the end after successfully retrieving.
- 12D3KooWDCXxiSsLi1NT9tsiyimwV6YstQkrjTjD2hAkz2KRVAGG
- QmUA9D3H7HeCYsirB3KmPSvZh3dNXMZas6Lwgr4fv1HTTp
- 12D3KooWPgBdZSbmKbD7ZQGjU7gZCcCKGvWSnBf1q4xAbpDdtJaJ

It also fails to tell me which provider it successfully fetch from
```
Fetched [bafybeic56z3yccnla3cutmvqsn5zy3g24muupcsjtoyp3pu5pm5amurjx4] from []:
```

--- a/pkg/retriever/bitswapretriever.go
+++ b/pkg/retriever/bitswapretriever.go
@@ -3,6 +3,7 @@ package retriever
 import (
 	"bytes"
 	"context"
+	"errors"
 	"fmt"
 	"io"
 	"math"
@@ -242,8 +243,14 @@ func (br *bitswapRetrieval) RetrieveFromAsyncCandidates(ayncCandidates types.Inb
 	br.routing.RemoveProviders(br.request.RetrievalID)
 	br.bstore.RemoveLinkSystem(br.request.RetrievalID)
 	if err != nil {
-		// record failure
-		br.events(events.FailedRetrieval(br.clock.Now(), br.request.RetrievalID, bitswapCandidate, multicodec.TransportBitswap, err.Error()))
+		// If the local context was canceled and the parent wasn't, it could be from a block timeout
+		// (see lastBytesReceivedTimer), in which case it needs to be recorded as a failure.
+		// However, if the parent context has an error we want to exclude cancellation errors
+		// since the failure could be because another protocol finished, not because of a bitswap
+		// problem.
+		if br.ctx.Err() == nil || !errors.Is(err, context.Canceled) {
+			br.events(events.FailedRetrieval(br.clock.Now(), br.request.RetrievalID, bitswapCandidate, multicodec.TransportBitswap, err.Error()))
+		}
 		return nil, err
 	}
 	duration := br.clock.Since(startTime)

--- a/pkg/retriever/parallelpeerretriever.go
+++ b/pkg/retriever/parallelpeerretriever.go
@@ -327,7 +327,9 @@ func (retrieval *retrieval) runRetrievalCandidate(
 	// Setup in parallel
 	connectTime, err := retrieval.Protocol.Connect(connectCtx, retrieval, startTime, candidate)
 	if err != nil {
-		if ctx.Err() == nil { // not cancelled, maybe timed out though
+		// Exclude the case where the context was cancelled by the parent, whichikely means that
+		// another protocol has succeeded.
+		if ctx.Err() == nil || !errors.Is(err, context.Canceled) {
 			logger.Warnf("Failed to connect to SP %s on protocol %s: %v", candidate.MinerPeer.ID, retrieval.Protocol.Code().String(), err)
 			retrievalErr = fmt.Errorf("%w: %v", ErrConnectFailed, err)
 			if err := retrieval.Session.RecordFailure(retrieval.request.RetrievalID, candidate.MinerPeer.ID); err != nil {
@@ -347,13 +349,17 @@ func (retrieval *retrieval) runRetrievalCandidate(
 			stats, retrievalErr = retrieval.Protocol.Retrieve(ctx, retrieval, shared, timeout, candidate)
 
 			if retrievalErr != nil {
-				msg := retrievalErr.Error()
-				if errors.Is(retrievalErr, ErrRetrievalTimedOut) {
-					msg = fmt.Sprintf("timeout after %s", timeout)
-				}
-				shared.sendEvent(ctx, events.FailedRetrieval(retrieval.parallelPeerRetriever.Clock.Now(), retrieval.request.RetrievalID, candidate, retrieval.Protocol.Code(), msg))
-				if err := retrieval.Session.RecordFailure(retrieval.request.RetrievalID, candidate.MinerPeer.ID); err != nil {
-					logger.Errorf("Error recording retrieval failure for protocol %s: %v", retrieval.Protocol.Code().String(), err)
+				// Exclude the case where the context was cancelled by the parent, which likely
+				// means that another protocol has succeeded.
+				if ctx.Err() == nil || !errors.Is(err, context.Canceled) {
+					msg := retrievalErr.Error()
+					if errors.Is(retrievalErr, ErrRetrievalTimedOut) {
+						msg = fmt.Sprintf("timeout after %s", timeout)
+					}
+					shared.sendEvent(ctx, events.FailedRetrieval(retrieval.parallelPeerRetriever.Clock.Now(), retrieval.request.RetrievalID, candidate, retrieval.Protocol.Code(), msg))
+					if err := retrieval.Session.RecordFailure(retrieval.request.RetrievalID, candidate.MinerPeer.ID); err != nil {
+						logger.Errorf("Error recording retrieval failure for protocol %s: %v", retrieval.Protocol.Code().String(), err)
+					}
 				}
 			} else {
 				shared.sendEvent(ctx, events.Success(

